From: Rouven Czerwinski <r.czerwinski@pengutronix.de>
Date: Fri, 3 Dec 2021 15:01:10 +0100
Subject: [PATCH] ARM: imx: add state backend for i.MX8MP

Store the state on the eMMC. We remove the unneeded barebox partition,
since barebox lives in the boot partition of the eMMC.

Signed-off-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>
---
 arch/arm/dts/imx8mp-evk.dts | 70 +++++++++++++++++++++++++++++++++++++++++----
 1 file changed, 64 insertions(+), 6 deletions(-)

diff --git a/arch/arm/dts/imx8mp-evk.dts b/arch/arm/dts/imx8mp-evk.dts
index 3264ade4b877..45bee810dd1d 100644
--- a/arch/arm/dts/imx8mp-evk.dts
+++ b/arch/arm/dts/imx8mp-evk.dts
@@ -23,11 +23,69 @@
 		};
 	};
 
+	aliases {
+		state = &state_emmc;
+	};
+
 	gpio-leds {
 		status {
 			barebox,default-trigger = "heartbeat";
 		};
 	};
+
+	state_emmc: state {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "barebox,state";
+		magic = <0x98335604>;
+		backend-type = "raw";
+		backend = <&backend_state_emmc>;
+		backend-stridesize = <0x200>;
+
+		bootstate {
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			system0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				remaining_attempts@0 {
+					reg = <0x0 0x4>;
+					type = "uint32";
+					default = <3>;
+				};
+
+				priority@4 {
+					reg = <0x4 0x4>;
+					type = "uint32";
+					default = <20>;
+				};
+			};
+
+			system1 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				remaining_attempts@8 {
+					reg = <0x8 0x4>;
+					type = "uint32";
+					default = <3>;
+				};
+
+				priority@c {
+					reg = <0xc 0x4>;
+					type = "uint32";
+					default = <21>;
+				};
+			};
+
+			last_chosen@10 {
+				reg = <0x10 0x4>;
+				type = "uint32";
+			};
+		};
+	};
 };
 
 &ethphy1 {
@@ -54,14 +112,14 @@
 	#address-cells = <1>;
 	#size-cells = <1>;
 
-	partition@0 {
-		label = "barebox";
-		reg = <0x0 0xe0000>;
-	};
-
 	partition@e0000 {
 		label = "barebox-environment";
-		reg = <0xe0000 0x20000>;
+		reg = <0x0 0x20000>;
+	};
+
+	backend_state_emmc: partition@200000 {
+			label = "barebox-state";
+			reg = <0x200000 0xe00000>;
 	};
 };
 
